components/core/src/clp/GlobalMetadataDBConfig.cpp

```diff
@@ -1,110 +1,99 @@
-}
+namespace clp {
+// Constants
+constexpr int cMinPort{1};
+constexpr int cMaxPort{65'535};
+
+auto operator>>(std::istream& in, GlobalMetadataDBConfig::MetadataDBType& metadata_db_type)
+        -> std::istream& {
+    string db_type_string;
+    in >> db_type_string;
+
+    if ("sqlite" == db_type_string) {
+        metadata_db_type = GlobalMetadataDBConfig::MetadataDBType::SQLite;
+    } else if ("mysql" == db_type_string) {
+        metadata_db_type = GlobalMetadataDBConfig::MetadataDBType::MySQL;
+    } else {
+        throw invalid_argument("Unknown database type: " + db_type_string);
+    }
 
-static exception
-get_yaml_unconvertable_value_exception(string const& key_name, string const& destination_type) {
-    throw invalid_argument(
-            fmt::format("'{}' could not be converted to type '{}'", key_name, destination_type)
-    );
+    return in;
 }
 
-namespace clp {
-void GlobalMetadataDBConfig::parse_config_file(string const& config_file_path) {
-    YAML::Node config = YAML::LoadFile(config_file_path);
+auto GlobalMetadataDBConfig::add_command_line_options(
+        boost::program_options::options_description& options_description
+) -> void {
+    options_description.add_options()
+            ("db-type",
+             boost::program_options::value<MetadataDBType>(&m_metadata_db_type)->default_value(cDefaultMetadataDbType, cDefaultMetadataDbTypeName.data()),
+             "Database type [sqlite | mysql]")
+            ("db-host",
+             boost::program_options::value<string>(&m_metadata_db_host)->default_value(cDefaultMetadataDbHost.data()),
+             "[db-type=mysql] Database host")
+            ("db-port",
+             boost::program_options::value<int>(&m_metadata_db_port)->default_value(cDefaultMetadataDbPort),
+             "[db-type=mysql] Database port")
+            ("db-name",
+             boost::program_options::value<string>(&m_metadata_db_name)->default_value(cDefaultMetadataDbName.data()),
+             "[db-type=mysql] Database name")
+            ("db-table-prefix",
+             boost::program_options::value<string>(&m_metadata_table_prefix)->default_value(cDefaultMetadataTablePrefix.data()),
+             "[db-type=mysql] Database table prefix");
+}
 
-    if (!config["type"]) {
-        throw get_yaml_missing_key_exception("type");
+auto GlobalMetadataDBConfig::read_credentials_from_env() -> void {
+    auto const* db_username{std::getenv("CLP_DB_USER")};
+    if (nullptr != db_username) {
+        m_metadata_db_username = db_username;
     }
 
-    auto db_type_string = config["type"].as<string>();
-    if ("sqlite" == db_type_string) {
-        m_metadata_db_type = MetadataDBType::SQLite;
-    } else if ("mysql" == db_type_string) {
-        m_metadata_db_type = MetadataDBType::MySQL;
+    auto const* db_password{std::getenv("CLP_DB_PASS")};
+    if (nullptr != db_password) {
+        m_metadata_db_password = db_password;
+    }
+}
```

Comment:
should we force a password to be provided if read_credentials_from_env is called? (and only call read_credentials_from_env if the database type is mysql).

one concern is that if there's a typo in the envvar name, clp will just not use any password for the db, wonder what would be the error message from clp look like if that happens.

Edit: nvm now I see that they are validated in validate(). I still kind of feel it would be the best if we validate the credentials to be not empty in the read_credentials_from_env(), how do you think?

Reply:
> should we force a password to be provided if read_credentials_from_env is called?

it seems mysql does allow empty password (and even usernames? i haven't verified) so checking emptiness of such can be tricky. i didn't add such checks for now

> and only call read_credentials_from_env if the database type is mysql ... I still kind of feel it would be the best if we validate the credentials to be not empty in the read_credentials_from_env(), how do you think?

sure

Reply:
> > should we force a password to be provided if read_credentials_from_env is called?
> 
> it seems mysql does allow empty password (and even usernames? i haven't verified) so checking emptiness of such can be tricky. i didn't add such checks for now
> 
> > and only call read_credentials_from_env if the database type is mysql ... I still kind of feel it would be the best if we validate the credentials to be not empty in the read_credentials_from_env(), how do you think?
> 
> sure

True, then I guess we can skip the password?